platform/features/table/src/Table.js

```diff
@@ -0,0 +1,163 @@
+        }
+
+        /**
+         * Build column definitions based on supplied telemetry metadata
+         * @param metadata Metadata describing the domains and ranges available
+         * @returns {Table} This object
+         */
+        Table.prototype.buildColumns = function(metadata) {
+            var self = this;
+
+            this.columns = [];
+
+            if (metadata) {
+                metadata.forEach(function (metadatum) {
+                    //Push domains first
+                    metadatum.domains.forEach(function (domainMetadata) {
+                        self.addColumn(new DomainColumn(domainMetadata, self.telemetryFormatter));
+                    });
+                    metadatum.ranges.forEach(function (rangeMetadata) {
+                        self.addColumn(new RangeColumn(rangeMetadata, self.telemetryFormatter));
+                    });
+                });
+            }
+            return this;
+        };
+
+        /**
+         * Add a column definition to this Table
+         * @param {RangeColumn | DomainColumn | NameColumn} column
+         * @param {Number} [index] Where the column should appear (will be
+         * affected by column filtering)
+         */
+        Table.prototype.addColumn = function (column, index) {
+            if (typeof index === 'undefined') {
+                this.columns.push(column);
+            } else {
+                this.columns.splice(index, 0, column);
+            }
+        };
+
+        /**
+         * @private
+         * @param column
+         * @returns {*|string}
+         */
+        Table.prototype.getColumnTitle = function (column) {
+                return column.getTitle();
+        };
+
+        /**
+         * Get a simple list of column titles
+         * @returns {Array} The titles of the columns
+         */
+        Table.prototype.getHeaders = function() {
+            var self = this;
+            return this.columns.map(function (column){
+                return self.getColumnTitle(column);
+            });
+        };
+
+        /**
+         * Retrieve and format values for a given telemetry datum.
+         * @param telemetryObject The object that the telemetry data is
+         * associated with
+         * @param datum The telemetry datum to retrieve values from
+         * @returns {Object} Key value pairs where the key is the column
+         * title, and the value is the formatted value from the provided datum.
+         */
+        Table.prototype.getRowValues = function(telemetryObject, datum) {
+            var self = this;
+            return this.columns.reduce(function(rowObject, column){
+                var columnTitle = self.getColumnTitle(column),
+                    columnValue = column.getValue(telemetryObject, datum);
+
+                if (columnValue !== undefined && columnValue.text === undefined){
+                    columnValue.text = '';
+                }
+                // Don't replace something with nothing.
```

Comment:
Yeah, this is kind of a subtle problem that I encountered. I decided that when dealing with tables composed of multiple telemetry points, domains and ranges with the same title should be shown in the same column. At the moment it is just based on Title, but I think that in future we could be a bit cleverer with this and only combine columns with the same title AND the same type [range or domain] AND the same formatting... or something. It's either that or don't combine them at all for now and just suffix them with the object name or something to distinguish them when they're displayed.

The reason for this logic is that you can have two telemetry points with domains or ranges that have the same Title (eg. Time) but different keys (time vs. timestamp). The data passed to the directive is keyed on title (because that's what is shown in the headers). So this can cause data loss if the the value for 'time' is overwritten by the value for 'timestamp' (which is undefined).